src/ore/src/tracing/optimizer.rs

```diff
@@ -0,0 +1,313 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Tracing utilities for explainable plans.
+
+use std::fmt::Debug;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::path::PathBuf;
+use std::sync::Mutex;
+
+use tracing::{span, subscriber};
+use tracing_subscriber::{field, layer};
+
+/// A tracing layer used to accumulate a sequence of explainable plans.
+#[allow(missing_debug_implementations)]
+pub struct PlanTrace<T> {
+    path: Mutex<PathBuf>,
+    times: Mutex<Vec<std::time::Instant>>,
+    entries: Mutex<Vec<TraceEntry<T>>>,
+}
+
+/// A struct created as a reflection of a [`trace_plan`] call.
+#[allow(missing_debug_implementations)]
+pub struct TraceEntry<T> {
+    /// The instant at which an entry was created.
+    ///
+    /// Used to impose global sorting when merging multiple `TraceEntry`
+    /// arrays in a single array.
+    pub instant: std::time::Instant,
+    /// The time it took to run this optimization step.
+    pub duration: std::time::Duration,
+    /// Ancestor chain of span names (root is first, parent is last).
+    pub path: String,
+    /// The plan produced this step.
+    pub plan: T,
+}
+
+/// Trace a fragment of type `T` to be emitted as part of an `EXPLAIN OPTIMIZER
+/// TRACE` output.
+///
+/// For best compatibility with the existing UI (which at the moment is the only
+/// sane way to look at such `EXPLAIN` traces), code instrumentation should
+/// adhere to the following constraints:
+///
+/// 1.  The plan type should be listed in the layers created in the
+///     `OptimizerTrace` constructor.
+/// 2.  Each `trace_plan` should be unique within it's enclosing span and should
+///     represent the result of the stage idenified by that span. In particular,
+///     this means that functions that call `trace_plan` more than once need to
+///     construct ad-hoc spans (see the iteration spans in the `Fixpoint`
+///     transform for example).
+///
+/// As a consequence of the second constraint, a sequence of paths such as
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// ```
+/// is not well-formed as it is missing the results of the prefix paths at the
+/// end:
+/// ```text
+/// optimizer.foo.bar
+/// optimizer.foo.baz
+/// optimizer.foo
+/// optimizer
+/// ```
+///
+/// Also, note that full paths can be repeated within a pipeline, but adjacent
+/// duplicates are interpreted as separete invocations. For example, the
+/// sub-sequence
+/// ```text
+/// ... // preceding stages
+/// optimizer.foo.bar // 1st call
+/// optimizer.foo.bar // 2nd call
+/// ... // following stages
+/// ```
+/// will be rendered by the UI as the following tree structure.
+/// ```text
+/// optimizer
+///   ... // following stages
+///   foo
+///     bar // 2nd call
+///     bar // 1st call
+///   ... // preceding stages
+/// ```
+pub fn trace_plan<T: Clone + 'static>(plan: &T) {
```

Comment:
Conceptually, this should just attach an event of type `T` to the current span. However, since typed events are not yet supported, I implemented this a bit differently. Hopefully the implementation can be simplified once `Valuable` is stable.

Reply:
I actually like what you have done here with the `PlanTrace<T>` layers, but I am curious, once we have valuable, can we remove this function, and just have something like `tracing::trace!(plan.as_value())`? Needing to do this `with_subscriber` stuff makes sense for how you implemented it, but avoiding that in the future seems cleaner

Reply:
>  can we remove this function, and just have something like `tracing::trace!(plan.as_value())?`

I would love if this is the case, but this kind of depends on the `Valuable` API. If `Valuable` can be automatically derived for types that implement `serde::ser::Serialize` this will be pretty simple, otherwise we will need to add `Valuable` implementations (possibly derived ones) for all plan types in all IR types used in the optimizer and I'm not sure it will be worth carrying this code around for the sole purpose to simplify the `PlanTrace` and `OptimizerTrace` implementations.